pallets/pallet-deposit-storage/src/fungible/mod.rs

```diff
@@ -0,0 +1,238 @@
+pub struct PalletDepositStorageReason<Namespace, Key> {
+	pub(crate) namespace: Namespace,
+	pub(crate) key: Key,
+}
+
+impl<Namespace, Key> From<PalletDepositStorageReason<Namespace, Key>> for HoldReason {
+	fn from(_value: PalletDepositStorageReason<Namespace, Key>) -> Self {
+		// All the deposits ever taken like this will count towards the same hold
+		// reason.
+		Self::Deposit
+	}
+}
+
+impl<T> InspectHold<T::AccountId> for Pallet<T>
+where
+	T: Config,
+{
+	type Reason = PalletDepositStorageReason<T::Namespace, DepositKeyOf<T>>;
+
+	fn total_balance_on_hold(who: &T::AccountId) -> Self::Balance {
+		<T::Currency as InspectHold<T::AccountId>>::total_balance_on_hold(who)
+	}
+
+	// We return the balance of hold of a specific (namespace, key), which prevents
+	// mistakes since everything in the end is converted to `HoldReason`.
+	fn balance_on_hold(reason: &Self::Reason, who: &T::AccountId) -> Self::Balance {
+		let Some(deposit_entry) = SystemDeposits::<T>::get(&reason.namespace, &reason.key) else {
+			return Zero::zero();
+		};
+		if deposit_entry.deposit.owner == *who {
+			deposit_entry.deposit.amount
+		} else {
+			Zero::zero()
+		}
+	}
+}
+
+// This trait is implemented by forwarding everything to the `Currency`
+// implementation.
+impl<T> Unbalanced<T::AccountId> for Pallet<T>
+where
+	T: Config,
+{
+	fn handle_dust(dust: Dust<T::AccountId, Self>) {
+		<T::Currency as Unbalanced<T::AccountId>>::handle_dust(Dust(dust.0));
+	}
+
+	fn write_balance(who: &T::AccountId, amount: Self::Balance) -> Result<Option<Self::Balance>, DispatchError> {
+		<T::Currency as Unbalanced<T::AccountId>>::write_balance(who, amount)
+	}
+
+	fn set_total_issuance(amount: Self::Balance) {
+		<T::Currency as Unbalanced<T::AccountId>>::set_total_issuance(amount);
+	}
+}
+
+impl<T> UnbalancedHold<T::AccountId> for Pallet<T>
+where
+	T: Config,
+{
+	/// This function gets called inside all invocations of `hold`, `release`,
+	/// and `release_all`. The value of `amount` is always the final amount that
+	/// should be kept on hold, meaning that a `release_all` will result in a
+	/// value of `0`. An hold increase from `1` to `2` will imply a value of
+	/// `amount` of `2`, i.e., the total amount held at the end of the
+	/// evaluation. Similarly, a decrease from `3` to `1` will imply a `amount`
+	/// value of `1`.
+	fn set_balance_on_hold(reason: &Self::Reason, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		SystemDeposits::<T>::try_mutate_exists(&reason.namespace, &reason.key, |maybe_existing_deposit_entry| {
+			match maybe_existing_deposit_entry {
+				// If this is the first registered deposit for this reason, create a new storage entry.
+				None => {
+					if amount > Zero::zero() {
+						// Increase the held amount for the final runtime hold reason by `amount`.
+						<T::Currency as UnbalancedHold<T::AccountId>>::increase_balance_on_hold(
+							&T::RuntimeHoldReason::from(reason.clone().into()),
```

Comment:
I see multiple clone().into() 

can we maybe move that into a variable?

Reply:
Good catch! https://github.com/KILTprotocol/kilt-node/pull/823/commits/44d9745f80725462d4ca86e1d98692879448e775